### ISSUE TITLE
FileSystemWatcher - Unhandled exception, taking down whole application

### ISSUE THREAD

**Issue by IonKiwi** [dotnet/runtime#41737]
### Description

We are using a file wachter, to detect changes to a directory.

`PhysicalFileProvider.Watch("/**");`

We expect the watcher to not throw any exceptions.

When the whole directory is removed, it throws an unhandled exception, and the container is killed.

The wachter should not throw any exception, killing the process.

Similar issue: https://github.com/dotnet/extensions/issues/1902

### Configuration

Running inside a docker ASP.NET Core 3.1.7-alpine Linux container.

### Other information

Stacktrace:

```
Unhandled exception. System.IO.DirectoryNotFoundException: Could not find a part of the path '/storage/Views'.
   at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
   at System.IO.Enumeration.FileSystemEnumerator`1.Init()
   at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
   at System.IO.Enumeration.FileSystemEnumerableFactory.FileSystemInfos(String directory, String expression, EnumerationOptions options, Boolean isNormalized)
   at System.IO.DirectoryInfo.InternalEnumerateInfos(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
   at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext()
   at System.Linq.Enumerable.OfTypeIterator[TResult](IEnumerable source)+MoveNext()
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath)
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute()
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges()
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged()
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state)

```

**Comment by carlossanlop** [comment]
Thanks for reporting @IonKiwi . Can you share a small code snippet with a repro?

**Comment by IonKiwi** [comment]
Seems to consistently repro within WSL2

Create a new WebApplication add/create PhysicalFileProvider and Watch in Configure:

```
		public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerFactory loggerFactory) {

			var logger = loggerFactory.CreateLogger<Startup>();

			var fileProvider = new PhysicalFileProvider(@"/data/views");
			var changeToken = fileProvider.Watch("/**");
			changeToken.RegisterChangeCallback((state) => { logger.LogWarning("Changed"); }, null);

...

			app.UseEndpoints(endpoints => {
				endpoints.MapControllerRoute(
						name: "default",
						pattern: "{controller=Home}/{action=Index}/{id?}");
			});
		}
```

Existing directory with content on the host
C:\data\views

Run the Docker container, and map C:\data to /data in the container
```
docker run -d -e "DOTNET_USE_POLLING_FILE_WATCHER=1" -p 32769:80 --mount type=bind,src=/mnt/c/data,dst=/data,ro --name repro webapplication2:latest
```

Modify and remove directory on the host with PowerShell
```
cd C:\data
echo 1 > views\test.txt
Remove-Item views -recurse -force
```

Container crashes
```
docker logs -f -t repro
2020-09-02T19:10:25.654391600Z warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
2020-09-02T19:10:25.654448000Z       Storing keys in a directory '/root/.aspnet/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
2020-09-02T19:10:25.680993100Z warn: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[35]
2020-09-02T19:10:25.681022000Z       No XML encryptor configured. Key {15bf0a12-bbb2-448b-a262-ba3ea5bc5fd4} may be persisted to storage in unencrypted form.
2020-09-02T19:10:25.862508900Z info: Microsoft.Hosting.Lifetime[0]
2020-09-02T19:10:25.862538100Z       Now listening on: http://[::]:80
2020-09-02T19:10:25.863638500Z info: Microsoft.Hosting.Lifetime[0]
2020-09-02T19:10:25.863651700Z       Application started. Press Ctrl+C to shut down.
2020-09-02T19:10:25.863656200Z info: Microsoft.Hosting.Lifetime[0]
2020-09-02T19:10:25.863660300Z       Hosting environment: Production
2020-09-02T19:10:25.863663900Z info: Microsoft.Hosting.Lifetime[0]
2020-09-02T19:10:25.863667700Z       Content root path: /app
2020-09-02T19:10:41.770823700Z Unhandled exception. System.IO.DirectoryNotFoundException: Could not find a part of the path '/data/views'.
2020-09-02T19:10:41.770989100Z    at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
2020-09-02T19:10:41.771036600Z    at System.IO.Enumeration.FileSystemEnumerator`1.Init()
2020-09-02T19:10:41.771049000Z    at System.IO.Enumeration.FileSystemEnumerator`1..ctor(String directory, Boolean isNormalized, EnumerationOptions options)
2020-09-02T19:10:41.771058000Z    at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
2020-09-02T19:10:41.771066500Z    at System.IO.Enumeration.FileSystemEnumerableFactory.FileSystemInfos(String directory, String expression, EnumerationOptions options, Boolean isNormalized)
2020-09-02T19:10:41.771074400Z    at System.IO.DirectoryInfo.InternalEnumerateInfos(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
2020-09-02T19:10:41.771082200Z    at System.IO.DirectoryInfo.EnumerateFileSystemInfos(String searchPattern, SearchOption searchOption)
2020-09-02T19:10:41.771089800Z    at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext()
2020-09-02T19:10:41.771097500Z    at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
2020-09-02T19:10:41.771105500Z    at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
2020-09-02T19:10:41.771113800Z    at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath)
2020-09-02T19:10:41.771122400Z    at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute()
2020-09-02T19:10:41.771149600Z    at Microsoft.Extensions.FileSystemGlobbing.Matcher.Execute(DirectoryInfoBase directoryInfo)
2020-09-02T19:10:41.771177800Z    at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges()
2020-09-02T19:10:41.771186200Z    at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged()
2020-09-02T19:10:41.771194300Z    at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state)
2020-09-02T19:10:41.771203500Z    at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool)
2020-09-02T19:10:41.771213500Z    at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
2020-09-02T19:10:41.771221600Z    at System.Threading.TimerQueue.FireNextTimers()
```

**Comment by IonKiwi** [comment]
More concrete repro with a console app.

```
static async Task Main(string[] args) {
	using (var root = new DisposableFileSystem()) {

		PhysicalFileProvider pfp = new PhysicalFileProvider(root.RootPath) {
			UsePollingFileWatcher = true,
			UseActivePolling = true,
		};
		var changeToken = pfp.Watch("**");
		changeToken.RegisterChangeCallback(o => { Console.WriteLine("test"); }, null);

		var pollingChangeToken = (PollingWildCardChangeToken)((CompositeChangeToken)changeToken).ChangeTokens[1];

		Directory.Delete(root.RootPath, true);
		await Task.Delay(5000); // wait for the pollingChangeToken.PollingInterval = {00:00:04}

		bool hasChanged = pollingChangeToken.HasChanged;
	}
}
```

Results in the same exception

```
Unhandled exception. System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\ewout\AppData\Local\Temp\yhiunpie.5ul'.
   at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
   at System.IO.Enumeration.FileSystemEnumerator`1.Init()
   at System.IO.Enumeration.FileSystemEnumerator`1..ctor(String directory, Boolean isNormalized, EnumerationOptions options)
   at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
   at System.IO.Enumeration.FileSystemEnumerableFactory.FileSystemInfos(String directory, String expression, EnumerationOptions options, Boolean isNormalized)
   at System.IO.DirectoryInfo.InternalEnumerateInfos(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
   at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath)
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute()
   at Microsoft.Extensions.FileSystemGlobbing.Matcher.Execute(DirectoryInfoBase directoryInfo)
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges()
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged()
   at ConsoleApp2.Program.Main(String[] args) in C:\dev\ConsoleApp2\ConsoleApp2\Program.cs:line 25
   at ConsoleApp2.Program.<Main>(String[] args)
```

**Comment by IonKiwi** [comment]
We also see, other IOExceptions bringing the container down when using the wachter.
Any chance this will get fixed in 3.1?

```
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://+:80'. Binding to endpoints defined in UseKestrel() instead.
Unhandled exception. System.IO.IOException: I/O error
   at System.IO.FileStatus.EnsureStatInitialized(ReadOnlySpan`1 path, Boolean continueOnError)
   at System.IO.FileSystemInfo.Init(FileStatus& fileStatus)
   at System.IO.FileSystemInfo.Create(String fullPath, String fileName, FileStatus& fileStatus)
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
   at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext()
   at System.Linq.Enumerable.OfTypeIterator[TResult](IEnumerable source)+MoveNext()
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath)
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute()
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges()
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged()
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state)
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool)
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
   at System.Threading.TimerQueue.FireNextTimers()
```

**Comment by IonKiwi** [comment]
Seems to be fixed in master with
https://github.com/dotnet/runtime/pull/44671 / https://github.com/dotnet/runtime/issues/44652

Deleting the directory does not crash the whole process anymore
